### PR TITLE
`azurerm_web_application_firewall_policy`: ruleset support `Microsoft_DefaultRuleSet_2.1`

### LINKED ISSUE #23802: Support for Regional WAF Default Rule Set 2.1
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

Hi,

DRS 2.1 is now general available for Regional Web Application Firewall Policies.

Could you allow the corresponding configuration in the provider?

Thanks :)

### New or Affected Resource(s)/Data Source(s)

azurerm_web_application_firewall_policy

### Potential Terraform Configuration

```hcl
managed_rule_set {
      type    = "DRS"
      version = "2.1"
}
```

### References

- [Azure GA Notification](https://azure.microsoft.com/en-us/updates/general-availability-default-rule-set-21-for-regional-waf-with-application-gateway/)

- [MS Docs](https://learn.microsoft.com/en-us/azure/web-application-firewall/ag/application-gateway-crs-rulegroups-rules?tabs=drs21#drs-21)

- https://github.com/hashicorp/terraform-provider-azurerm/blob/22ba1063bb1d5a094e513ee3139dbcf5b0aa61d5/internal/services/network/validate/web_application_firewall_policy.go#L40-L52

### PR DESCRIPTION
**Pull request by wuxu92**
fixes #23802

```
--- PASS: TestAccWebApplicationFirewallPolicy_ManagedRuleSetDRS (258.69s)
PASS
```

### CODE REVIEW

File: website/docs/r/web_application_firewall_policy.html.markdown [hunk 1]
```diff
@@ -216,9 +216,9 @@ The `exclusion` block supports the following:
 
 The `excluded_rule_set` block supports the following:
 
-* `type` - (Optional) The rule set type. The only possible value is `OWASP` . Defaults to `OWASP`.
+* `type` - (Optional) The rule set type. The only possible value is `Microsoft_DefaultRuleSet` and `OWASP`. Defaults to `OWASP`.
```
Comment (jackofallops) [inline website/docs/r/web_application_firewall_policy.html.markdown:219]:
```suggestion
* `type` - (Optional) The rule set type. Possible values include `Microsoft_DefaultRuleSet` and `OWASP`. Defaults to `OWASP`.
```

Comment (wuxu92) [reply website/docs/r/web_application_firewall_policy.html.markdown]: updated.

File: website/docs/r/web_application_firewall_policy.html.markdown [hunk 2]
```diff
@@ -216,9 +216,9 @@ The `exclusion` block supports the following:
...
 
-* `version` - (Optional) The rule set version. The only possible value is `3.2` . Defaults to `3.2`.
+* `version` - (Optional) The rule set version. The only possible value is `2.1` (for rule set type `Microsoft_DefaultRuleSet`) and `3.2` (for rule set type `OWASP`). Defaults to `3.2`.
```
[the remark this answers is not in the document]
Comment (wuxu92) [reply website/docs/r/web_application_firewall_policy.html.markdown]: updated.

File: internal/services/network/validate/web_application_firewall_policy.go [hunk 3]
```diff
@@ -49,12 +66,15 @@ var ValidateWebApplicationFirewallPolicyRuleSetVersion = validation.StringInSlic
 var ValidateWebApplicationFirewallPolicyRuleSetType = validation.StringInSlice([]string{
```
Comment (teowa) [inline internal/services/network/validate/web_application_firewall_policy.go:66]: Hi Xu, this validation function is also used by `azurerm_application_gateway`, however the field in `azurerm_application_gateway` don't support the new value, with below error:
```
│ Error: creating/updating "Resource: (ResourceId \"/subscriptions/xxx/resourceGroups/wt-acctestRG/providers/Microsoft.Network/applicationGateways/testapgw\" / Api Version \"2023-06-01\")": PUT https://management.azure.com/subscriptions/xxx/resourceGroups/wt-acctestRG/providers/Microsoft.Network/applicationGateways/testapgw
│ --------------------------------------------------------------------------------
│ RESPONSE 400: 400 Bad Request
│ ERROR CODE: ApplicationGatewayFirewallUnknownRuleSetType
│ --------------------------------------------------------------------------------
│ {
│   "error": {
│     "code": "ApplicationGatewayFirewallUnknownRuleSetType",
│     "message": "RuleSetType 'Microsoft_DefaultRuleSet' does not exist for Application Gateway Firewall in context 'properties.webApplicationFirewallConfiguration'.",
│     "details": []
│   }
│ }
│ --------------------------------------------------------------------------------
│
```
So this validate function may need to be split. More details can be found in my duplicated PR https://github.com/hashicorp/terraform-provider-azurerm/pull/23899

Comment (teowa) [reply internal/services/network/validate/web_application_firewall_policy.go:66]: Seems will not be supported even in latest API `2023-06-01` as I have tested using azapi.

[the remark this answers is not in the document]
Comment (manicminer) [reply internal/services/network/validate/web_application_firewall_policy.go:66]: @wuxu92 Could you elaborate more on this? Will these changes work or does this now depend on an API/service update or fix? Thanks!

[the remark this answers is not in the document]
Comment (bubbletroubles) [reply internal/services/network/validate/web_application_firewall_policy.go:66]: The WAF policy it works for me with the following code using `azapi`

```
resource "azapi_resource" "example" {
  type      = "Microsoft.Network/ApplicationGatewayWebApplicationFirewallPolicies@2023-04-01"
  name      = "waf-name"
  location  = var.secondary_location
  parent_id = var.secondary_resource_group_id
  tags      = var.tags
  schema_validation_enabled = false
  body = jsonencode({
    properties = {
      customRules = [
        {
          action = "Block"
          matchConditions = [
            {
              matchValues = [
                "my-value"
              ]
              matchVariables = [
                {
                  variableName = "RequestHeaders"
                  selector     = "my-header"
                }
              ]
              negationConditon = true
              operator         = "Equal"
              transforms       = []
            }
          ]
          name = "blockrule"
          priority : 10
          ruleType : "MatchRule"
        },
        {
          action = "Allow"
          matchConditions = [
            {
              matchVariables = [
                {
                  variableName = "RequestHeaders"
                  selector     = "Content-type"
                }
              ]
              negationConditon = false
              operator         = "Contains"
              transforms       = []
              matchValues = [
                "application/cloudevents+json"
              ]
            },
            {
              matchVariables = [
                {
                  variableName = "RequestUri"
                }
              ]
              negationConditon = false
              operator         = "Contains"
              transforms       = []
              matchValues = [
                "my/uri"
              ]
            }
          ]
          name = "AllowMyApp"
          priority : 100
          ruleType : "MatchRule"
        }
      ]
      managedRules = {
        exclusions = []
        managedRuleSets = [
          {
            ruleSetType        = "Microsoft_DefaultRuleSet"
            ruleSetVersion     = "2.1"
            ruleGroupOverrides = []
          }
        ]
      }
      policySettings = {
        fileUploadLimitInMb    = 100
        maxRequestBodySizeInKb = 128
        state                  = "Enabled"
        mode                   = "Prevention"
        requestBodyCheck       = true
      }
    }
  })
}
```

**Review by jackofallops — APPROVED** [review]
Thanks for the changes @wuxu92 - This LGTM now 👍

**Pull request merged**